src/Tools/ExternalAccess/Extensions/Internal/ExtensionWorspaceMessageHandlerWrapper.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Extensions;
+
+internal sealed class ExtensionWorspaceMessageHandlerWrapper : IExtensionWorspaceMessageHandlerWrapper
+{
+    private readonly object handler;
+    private readonly MethodInfo executeAsyncMethod;
+    private readonly PropertyInfo responseTaskResultProperty;
+
+    public ExtensionWorspaceMessageHandlerWrapper(object handler, Type iCustomMessageHandlerInterface)
+    {
+        this.handler = handler;
+
+        Name = handler.GetType().FullName;
+        MessageType = iCustomMessageHandlerInterface.GenericTypeArguments[0];
+        ResponseType = iCustomMessageHandlerInterface.GenericTypeArguments[1];
+
+        executeAsyncMethod = iCustomMessageHandlerInterface.GetMethod(nameof(IExtensionWorkspaceMessageHandler<,>.ExecuteAsync));
+        responseTaskResultProperty = typeof(Task<>).MakeGenericType(ResponseType).GetProperty(nameof(Task<>.Result));
+    }
+
+    public Type MessageType { get; }
+
+    public Type ResponseType { get; }
+
+    public string Name { get; }
+
+    public async Task<object?> ExecuteAsync(object? message, Solution solution, CancellationToken cancellationToken)
+    {
+        if ((message is null && MessageType.IsValueType) || (message is not null && !MessageType.IsAssignableFrom(message.GetType())))
+        {
+            throw new InvalidOperationException($"The message type {message?.GetType().FullName ?? "null"} is not assignable to {MessageType.FullName}.");
+        }
+
+        var responseTask = (Task)executeAsyncMethod.Invoke(handler, [message, new ExtensionMessageContext(solution), cancellationToken]);
+        await responseTask.ConfigureAwait(false);
+        var response = responseTaskResultProperty.GetValue(responseTask);
+
+        if ((response is null && ResponseType.IsValueType) || (response is not null && !ResponseType.IsAssignableFrom(response.GetType())))
+        {
+            throw new InvalidOperationException($"The message type {response?.GetType().FullName ?? "null"} is not assignable to {ResponseType.FullName}.");
+        }
+
+        return response;
+    }
```

Comment:
lots of duplicate code as well.  would really like to share.

Reply:
Done